Fixes #78, fully initialize nodeLib

ixmatus opened the pull request:
... to include the tools sub-directory.

This change adds the `Turtle` library, providing the useful `cptree` function.

Note that `Turtle.cptree` clobbers so this does not preserve the original behavior of _not_ overwriting existing files within the generated `nodeLib` directory. My rationale as to why this is okay is as follows: (A) initialization occurs only once when the tool is run so it isn't expensive to copy the whole tree, (B) because it is generated there is no concern of overwriting manual modifications since users shouldn't manually modify generated code (they should instead override), and (C) rewriting the `nodeLib` on initialization ensures newer `nodeLib` Nix expressions are written if a new version of the utility is being executed.

adnelson commented on src/NixFromNpm/Conversion/ToDisk.hs:
```diff
@@ -179,11 +180,9 @@ initializeOutput = do
-          unlessExists (outputPath </> "nodeLib" </> filename path) $
-            copyFile path (outputPath </> "nodeLib" </> filename path)
+
+      Turtle.cptree nodeLibPath (outputPath </> "nodeLib")
```
We already have `cp_r` imported from `shelly`; is there a compelling reason to add a new dependency for this behavior?

adnelson replied:
I'm not super worried about the clobbering, but we might want to `rm` the previous `nodeLib` directory so that unused files are not left behind. Of course this could be a problem if people put their own custom files in there. Not sure what the best way to go about that is

ixmatus replied:
I wasn't aware of `cp_r`. I'll go ahead and use that to eliminate the extra dep.

I think removing `nodeLib` is fine.

adnelson commented:
cp_r works in the Sh monad so you'll have to wrap it with shelly. But yeah
I think it should work.

Honestly turtle is probably just fine (maybe better!) but it's one more
dependency, and in Haskell that usually means 50 more dependencies :P

On Thu, Aug 31, 2017 at 16:15 Parnell Springmeyer <notifications@github.com>
wrote:

> *@ixmatus* commented on this pull request.
> ------------------------------
>
> In src/NixFromNpm/Conversion/ToDisk.hs
> <https://github.com/adnelson/nixfromnpm/pull/79#discussion_r136450968>:
>
> > @@ -179,11 +180,9 @@ initializeOutput = do
>        -- Get the path to the files bundled with nixfromnpm which contain
>        -- nix libraries.
>        nodeLibPath <- (</> "nodeLib") <$> getDataFileName "nix-libs"
> -      -- Copy each of these into the target (unless they exist).
> -      forItemsInDir_ nodeLibPath $ \path -> do
> -        whenM (isFile path) $ do
> -          unlessExists (outputPath </> "nodeLib" </> filename path) $
> -            copyFile path (outputPath </> "nodeLib" </> filename path)
> +
> +      Turtle.cptree nodeLibPath (outputPath </> "nodeLib")
>
> I wasn't aware of cp_r. I'll go ahead and use that to eliminate the extra
> dep.
>
> I think removing nodeLib is fine.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/adnelson/nixfromnpm/pull/79#discussion_r136450968>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2wtzl36JEbkzo52DtnHCuVsSHQ4kkjks5sdyJagaJpZM4PJYSf>
> .
>
-- 
-- Allen

ixmatus commented:
I'm of the philosophy that my changes should be as consistent with the project as possible. `Turtle` is pretty light-weight but I think `Shelly` is just fine and it's used by this project so we should keep using it :)

I believe I addressed your concerns with 6f2c3bf, thanks for being so responsive @adnelson!

adnelson commented on src/NixFromNpm/Conversion/ToDisk.hs:
```diff
@@ -174,16 +174,17 @@ initializeOutput = do
-        whenM (isFile path) $ do
-          unlessExists (outputPath </> "nodeLib" </> filename path) $
-            copyFile path (outputPath </> "nodeLib" </> filename path)
+
```
Let's keep the comments in there

adnelson commented on src/NixFromNpm/Conversion/ToDisk.hs:
```diff
@@ -174,16 +174,17 @@ initializeOutput = do
       unlessExists defaultNixPath $
         writeNix (outputPath </> "default.nix") $ rootDefaultNix npm3
-      createDirectoryIfMissing (outputPath </> "nodeLib")
+      let outputNodeLib = outputPath </> "nodeLib"
```
I feel like this should be defined alongside its `input` counterpart.

ixmatus commented:
Don't merge this yet, I think shelly is not setting permissions correctly on the tree that is being created with `cp_r`.

ixmatus commented:
Actually, shelly is doing what it's supposed to be doing; it's that the tree coming from the package has read only permissions so those are being persisted to the copy location.

ixmatus commented:
Okay, I figured it out, the executable bit is not being preserved for the scripts under `nodeLib/tools`, I'll push another commit to this PR to fix that.

adnelson commented:
Awesome thanks! By the way I'm not at my computer so I won't be able to
test the changes myself until a little later.

On Thu, Aug 31, 2017 at 17:47 Parnell Springmeyer <notifications@github.com>
wrote:

> Okay, I figured it out, the executable bit is not being preserved for the
> scripts under nodeLib/tools, I'll push another commit to this PR to fix
> that.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/adnelson/nixfromnpm/pull/79#issuecomment-326440570>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE2wt4UtTe4WRzITPBFjq6sWautGA7Aqks5sdzgBgaJpZM4PJYSf>
> .
>
-- 
-- Allen

adnelson commented on src/NixFromNpm/Conversion/ToDisk.hs:
```diff
@@ -174,16 +175,24 @@ initializeOutput = do
+        cp_r inputNodeLib outputNodeLib
+
+        let tools     = outputNodeLib </> "tools"
+        let toolsText = either id id (toText tools)
```
have a function `pathToText` which does the same thing. It should already be in scope

adnelson commented on src/NixFromNpm/Conversion/ToDisk.hs:
```diff
@@ -174,16 +175,24 @@ initializeOutput = do
+        rm_rf outputNodeLib
+        cp_r inputNodeLib outputNodeLib
+
+        let tools     = outputNodeLib </> "tools"
```
while I don't have a problem with lining up the `=` in the `let`s, that's not the formatting used elsewhere in the file, so for consistency let's just use `let tools = outputNodeLib </> "tools"`

adnelson commented on src/NixFromNpm/Options.hs:
```diff
@@ -205,7 +205,7 @@ validateOptions opts@(RawOptions{..}) = do
   npmTokensEnv <- getNpmTokens
   return $ NixFromNpmOptions {
-    nfnoOutputPath = outputPath,
+    nfnoOutputPath = collapse outputPath,
```
I'm not familiar with the `collapse` function; what's going on here?

ixmatus replied:
[`Filesystem.Path.collapse`](https://hackage.haskell.org/package/system-filepath-0.4.13.4/docs/Filesystem-Path.html#v:collapse) will resolve the interstitial special directory path segments `.` and `..` to a collapsed absolute path.

```haskell
>>> :set -XOverloadedStrings
>>> import qualified Filesystem.Path
>>> import qualified Filesystem.Path.CurrentOS
>>> Filesystem.Path.collapse (Filesystem.Path.CurrentOS.fromText "/home/parnell/./Desktop/../")
FilePath "/home/parnell/"
```

Most tools will handle the special dirs but it's good practice I think to collapse them to the resolved absolute path when possible since some tools won't handle it well. I noticed that `nixfromnpm` wasn't collapsing dir paths when the Nix build script was failing to run `wrapProgram` on the scripts within the `tools/` directory. I had invoked `nixfromnpm` like this:

```shell
[nix-shell:~/Development/work]$ nixfromnpm --registry <private_registry> -p pulp -o ./
```

adnelson replied:
Ah this is called `normpath` in python. I agree it's a good practice. :+1:

adnelson commented:
This looks good! Thanks so much for the work on this one!

ixmatus commented:
Yas!
